diff -ur gsoap-2.8.orig/gsoap/wsdl/types.cpp gsoap-2.8/gsoap/wsdl/types.cpp
--- gsoap-2.8.orig/gsoap/wsdl/types.cpp	2018-01-15 19:48:15.000000000 +0100
+++ gsoap-2.8/gsoap/wsdl/types.cpp	2018-02-01 14:30:41.041321054 +0100
@@ -3240,7 +3240,7 @@
       if (cflag || sflag)
       {
         fprintf(stream, "/// Size of the dynamic array of %s is %s..%s.\n", s, min ? min : "1", max);
-        fprintf(stream, sizeformat, vname("$SIZE"), aname(NULL, NULL, name, &members));
+        fprintf(stream, sizeformat, vname("$SIZE"), aname(NULL, NULL, name));
         fprintf(stream, " %s", fake_union ? "0" : min ? min : "1");
         if (is_integer(max))
           fprintf(stream, ":%s", max);
@@ -3297,7 +3297,7 @@
       if (cflag || sflag)
       {
         fprintf(stream, "/// Size of array of %s is %s..%s.\n", s, min ? min : "1", max);
-        fprintf(stream, sizeformat, vname("$SIZE"), aname(NULL, NULL, name, &members));
+        fprintf(stream, sizeformat, vname("$SIZE"), aname(NULL, NULL, name));
         fprintf(stream, " %s", fake_union ? "0" : min ? min : "1");
         if (is_integer(max))
           fprintf(stream, ":%s", max);
@@ -3340,7 +3340,7 @@
       if (cflag || sflag)
       {
         fprintf(stream, "/// Size of %s array is %s..%s.\n", name, min ? min : "1", max);
-        fprintf(stream, sizeformat, vname("$SIZE"), aname(NULL, NULL, name, &members));
+        fprintf(stream, sizeformat, vname("$SIZE"), aname(NULL, NULL, name));
         fprintf(stream, " %s", fake_union ? "0" : min ? min : "1");
         if (is_integer(max))
           fprintf(stream, ":%s", max);
@@ -3393,7 +3393,7 @@
       if (cflag || sflag)
       {
         fprintf(stream, "/// Size of %s array is %s..%s.\n", name, min ? min : "1", max);
-        fprintf(stream, sizeformat, vname("$SIZE"), aname(NULL, NULL, name, &members));
+        fprintf(stream, sizeformat, vname("$SIZE"), aname(NULL, NULL, name));
         fprintf(stream, " %s", fake_union ? "0" : min ? min : "1");
         if (is_integer(max))
           fprintf(stream, ":%s", max);
@@ -3474,7 +3474,7 @@
     {
       if (cflag || sflag)
       {
-        fprintf(stream, sizeformat, vname("$SIZE"), aname(NULL, NULL, name, &members));
+        fprintf(stream, sizeformat, vname("$SIZE"), aname(NULL, NULL, name));
         fprintf(stream, " %s", fake_union ? "0" : min ? min : "1");
         if (is_integer(max))
           fprintf(stream, ":%s", max);
