diff -ur gsoap-2.8.91.orig/gsoap/doc/guide/stdsoap2.h gsoap-2.8.91/gsoap/doc/guide/stdsoap2.h
--- gsoap-2.8.91.orig/gsoap/doc/guide/stdsoap2.h	2019-08-15 14:05:04.000000000 +0200
+++ gsoap-2.8.91/gsoap/doc/guide/stdsoap2.h	2019-08-16 15:31:18.664273702 +0200
@@ -2730,7 +2730,7 @@
 struct soap *soap3 = soap_new2(input_mode, output_mode);
 ~~~
 
-Alternatively, use contructors in C++ as follows:
+Alternatively, use constructors in C++ as follows:
 
 ~~~{.cpp}
 struct soap *soap1();
@@ -2765,7 +2765,7 @@
 soap_init2(&soap3, input_mode, output_mode);
 ~~~
 
-Alternatively, use contructors in C++:
+Alternatively, use constructors in C++:
 
 ~~~{.cpp}
 struct soap soap1();
