diff --git a/magic/Magdir/convex b/magic/Magdir/convex
index 6141a82..4e096b9 100644
--- a/magic/Magdir/convex
+++ b/magic/Magdir/convex
@@ -30,8 +30,6 @@
 # The restore program uses these number to determine how the data is
 # to be extracted.
 #
-24	belong	=60011	dump format, 4.1 BSD or earlier
-24	belong	=60012	dump format, 4.2 or 4.3 BSD without IDC
 24	belong	=60013	dump format, 4.2 or 4.3 BSD (IDC compatible)
 24	belong	=60014	dump format, Convex Storage Manager by-reference dump
 #
