diff --git a/magic/Magdir/linux b/magic/Magdir/linux
index aa8d1aa..90f0f46 100644
--- a/magic/Magdir/linux
+++ b/magic/Magdir/linux
@@ -87,7 +87,17 @@
 >0x41a	beshort		x		\b%04x
 # From Daniel Novotny <dnovotny@redhat.com>
 # swap file for PowerPC
-65526	string		SWAPSPACE2	Linux/ppc swap file
+65526	string		SWAPSPACE2	Linux swap file
+>0x400  long        x       version %d,
+>0x404  long        x       size %d pages,
+>1052   string      \0      no label,
+>1052   string      >\0     LABEL=%s,
+>0x40c  belong      x       UUID=%08x
+>0x410  beshort     x       \b-%04x
+>0x412  beshort     x       \b-%04x
+>0x414  beshort     x       \b-%04x
+>0x416  belong      x       \b-%08x
+>0x41a  beshort     x       \b%04x
 16374   string          SWAPSPACE2      Linux/ia64 swap file
 #
 # Linux kernel boot images, from Albert Cahalan <acahalan@cs.uml.edu>
