diff --git a/magic/Magdir/elf b/magic/Magdir/elf
index e0e9937..11c876e 100644
--- a/magic/Magdir/elf
+++ b/magic/Magdir/elf
@@ -84,10 +84,21 @@
 >>>48	leshort		&0x0008		(LP64),
 >>18	leshort		16		nCUBE,
 >>18	leshort		17		Fujitsu VPP500,
->>18	leshort		18		SPARC32PLUS - invalid byte order,
->>18	leshort		20		PowerPC,
+>>18	leshort		18		SPARC32PLUS,
+# only for 32-bit
+>>>4	byte		1
+>>>>36	lelong&0xffff00	0x000100	V8+ Required,
+>>>>36	lelong&0xffff00	0x000200	Sun UltraSPARC1 Extensions Required,
+>>>>36	lelong&0xffff00	0x000400	HaL R1 Extensions Required,
+>>>>36	lelong&0xffff00	0x000800	Sun UltraSPARC3 Extensions Required,
+>>18	leshort		19		Intel 80960,
+>>18	leshort		20		PowerPC or cisco 4500,
+>>18	leshort		21		64-bit PowerPC or cisco 7500,
 >>18	leshort		22		IBM S/390,
->>18	leshort		36		NEC V800,
+>>18	leshort		23		Cell SPU,
+>>18	leshort		24		cisco SVIP,
+>>18	leshort		25		cisco 7200,
+>>18	leshort		36		NEC V800 or cisco 12000,
 >>18	leshort		37		Fujitsu FR20,
 >>18	leshort		38		TRW RH-32,
 >>18	leshort		39		Motorola RCE,
